File: system-test/typescript/src/testkit/dsl/core/scenario/execution-result-builder.ts
import type { Result } from '../../../common/result.js';
import type { SystemError } from '../../../driver/port/dtos/errors/SystemError.js';
import type { ScenarioContext } from './scenario-context.js';
import type { ExecutionResult } from './execution-result.js';

export class ExecutionResultBuilder<TResponse> {
  private _result: Result<TResponse, SystemError> | null = null;
  private _context: ScenarioContext | null = null;
  private _orderNumber: string | null = null;
  private _couponCode: string | null = null;

  withResult(result: Result<TResponse, SystemError>): this {
    this._result = result;
    return this;
  }

  withContext(context: ScenarioContext): this {
    this._context = context;
    return this;
  }

  orderNumber(orderNumber: string): this {
    this._orderNumber = orderNumber;
    return this;
  }

  couponCode(code: string): this {
    this._couponCode = code;
    return this;
  }

  getOrderNumber(): string | null {
    return this._orderNumber;
  }

  getCouponCode(): string | null {
    return this._couponCode;
  }

  build(): ExecutionResult<TResponse> {
    if (!this._result) {
      throw new Error('ExecutionResultBuilder: result is required');
    }
    if (!this._context) {
      throw new Error('ExecutionResultBuilder: context is required');
    }
    return { result: this._result, context: this._context };
  }
}

File: system/multitier/backend-java/src/testSupport/java/com/mycompany/myshop/backend/support/core/scenario/ExecutionResultBuilder.java
package com.mycompany.myshop.backend.support.core.scenario;

import com.mycompany.myshop.backend.support.core.shared.ResponseVerification;
import com.mycompany.myshop.backend.support.core.shared.UseCaseResult;

public class ExecutionResultBuilder<R, V extends ResponseVerification<R>> {

    private final UseCaseResult<R, V> result;
    private String orderNumber;
    private String couponCode;

    public ExecutionResultBuilder(UseCaseResult<R, V> result) {
        this.result = result;
    }

    public ExecutionResultBuilder<R, V> orderNumber(String orderNumber) {
        this.orderNumber = orderNumber;
        return this;
    }

    public ExecutionResultBuilder<R, V> couponCode(String couponCode) {
        this.couponCode = couponCode;
        return this;
    }

    public ExecutionResult<R, V> build() {
        return new ExecutionResult<>(result, orderNumber, couponCode);
    }
}
